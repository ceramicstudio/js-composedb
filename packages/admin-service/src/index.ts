export { router } from './router.js'
export type { Router } from './router.js'
export { GraphQLQueryCodec, Service } from './service.js'
export type { GraphQLResult, ServiceConfig, ServiceParams } from './service.js'
