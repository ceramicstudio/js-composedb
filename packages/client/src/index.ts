/**
 * High-level ComposeDB client, based on the {@linkcode runtime ComposeDB runtime}.
 *
 * ## Installation
 *
 * ```sh
 * npm install @composedb/client
 * ```
 *
 * @module client
 */

export { ComposeClient, type ComposeClientParams } from './client.js'
