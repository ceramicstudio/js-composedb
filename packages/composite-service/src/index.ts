export { router } from './router.js'
export type { CreateModelInput, LoadModelInput, SaveCompositeInput, Router } from './router.js'
export { Service } from './service.js'
export type { GraphQLResult, SavedComposite, ServiceParams } from './service.js'
