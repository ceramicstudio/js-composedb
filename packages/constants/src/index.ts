export const NODE_ID_FIELD = 'id'

export const VIEWER_ID_HEADER = 'ComposeDB-ViewerID'
