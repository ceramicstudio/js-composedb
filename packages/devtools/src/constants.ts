export const DOC_ID_FIELD = 'id'
