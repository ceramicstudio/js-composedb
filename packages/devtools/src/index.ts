/**
 * Development tools library.
 *
 * ## Installation
 *
 * ```sh
 * npm install --dev @composedb/devtools
 * ```
 *
 * @module devtools
 */

export * from './composite.js'
export * from './formats/json.js'
export { createRuntimeDefinition, getName } from './formats/runtime.js'
export { createAbstractCompositeDefinition } from './schema/compiler.js'
export * from './schema/types.js'
export * from './test-utils.js'
export * from './types.js'
export * from './utils.js'
