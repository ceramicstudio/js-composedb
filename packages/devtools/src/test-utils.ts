import type { ModelDefinition, ModelViewsDefinition } from '@ceramicnetwork/stream-model'
import type { InternalCompositeDefinition } from '@composedb/types'

import { createAbstractCompositeDefinition } from './schema/compiler.js'
import type { AbstractCompositeDefinition } from './schema/types.js'

/** @internal */
export function mockDefinition(
  definition: AbstractCompositeDefinition,
  providedModels: Record<string, ModelDefinition> = {}
): InternalCompositeDefinition {
  const models: Record<string, ModelDefinition> = {}
  const modelsViews: Record<string, ModelViewsDefinition> = {}

  for (const abstractModel of Object.values(definition.models)) {
    if (abstractModel.action === 'create') {
      const definition = abstractModel.model
      models[`${definition.name}ID`] = definition
    } else {
      const definition = providedModels[abstractModel.id]
      if (definition == null) {
        throw new Error(`Missing provided model ${abstractModel.id}`)
      }
      models[abstractModel.id] = definition
      modelsViews[abstractModel.id] = abstractModel.views
    }
  }

  return {
    version: '1.0',
    commonEmbeds: definition.commonEmbeds,
    models,
    views: { models: modelsViews },
  }
}

/** @internal */
export function mockDefinitionFromSchema(
  schema: string,
  providedModels?: Record<string, ModelDefinition>
): InternalCompositeDefinition {
  return mockDefinition(createAbstractCompositeDefinition(schema), providedModels)
}
