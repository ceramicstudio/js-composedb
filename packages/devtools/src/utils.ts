import type { ModelRelationViewDefinition, ModelViewDefinition } from '@ceramicnetwork/stream-model'
import type { StreamRef } from '@ceramicnetwork/streamid'
import type { RuntimeRelation, RuntimeRelationSource, RuntimeViewField } from '@composedb/types'

/** @internal */
export function streamIDToString(id: StreamRef | string): string {
  return typeof id === 'string' ? id : id.toString()
}

/** @internal */
export function applyMap<
  M extends Record<string, unknown>,
  V extends M[keyof M] = M[keyof M],
  R = unknown
>(inputs: M, callFunc: (input: V) => R): Record<keyof M, R> {
  return Object.entries(inputs).reduce((acc, [key, value]) => {
    acc[key as keyof M] = callFunc(value as V)
    return acc
  }, {} as Record<keyof M, R>)
}

/** @internal */
export async function promiseMap<
  M extends Record<string, unknown>,
  V extends M[keyof M] = M[keyof M],
  R = unknown
>(inputs: M, callFunc: (input: V) => Promise<R>): Promise<Record<keyof M, R>> {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
  const results = await Promise.all(Object.values(inputs).map((value) => callFunc(value as any)))
  return Object.keys(inputs).reduce((acc, key, i) => {
    acc[key as keyof M] = results[i]
    return acc
  }, {} as Record<keyof M, R>)
}

type RelationViewType = ModelRelationViewDefinition['type']

const RELATION_VIEW_SOURCES: Record<RelationViewType, RuntimeRelationSource> = {
  relationDocument: 'document',
  relationFrom: 'queryConnection',
  relationCountFrom: 'queryCount',
}

const RELATION_VIEW_TYPES = Object.keys(RELATION_VIEW_SOURCES)

/** @internal */
export function assertRelationViewDefinition(
  view: ModelViewDefinition
): ModelRelationViewDefinition {
  if (!RELATION_VIEW_TYPES.includes(view.type)) {
    throw new Error(`Invalid relation view: ${view.type}`)
  }
  return view as ModelRelationViewDefinition
}

/** @internal */
export function viewDefinitionToRuntimeRelation(view: ModelViewDefinition): RuntimeRelation {
  const relation = assertRelationViewDefinition(view)
  return {
    source: RELATION_VIEW_SOURCES[relation.type],
    model: relation.model,
    property: relation.property,
  }
}

/** @internal */
export function viewDefinitionToRuntime(view: ModelViewDefinition): RuntimeViewField {
  switch (view.type) {
    case 'documentAccount':
    case 'documentVersion':
      return { type: 'view', viewType: view.type }
    case 'relationCountFrom':
    case 'relationDocument':
    case 'relationFrom':
      return {
        type: 'view',
        viewType: 'relation',
        relation: viewDefinitionToRuntimeRelation(view),
      }
    default:
      // @ts-ignore unexpected view type
      throw new Error(`Unsupported view type: ${view.type as string}`)
  }
}

/** @internal */
export function viewRuntimeToModel(field: RuntimeViewField): ModelViewDefinition {
  if (field.viewType === 'relation') {
    const { model, property, source } = field.relation
    switch (source) {
      case 'document':
        return { type: 'relationDocument', model, property }
      case 'queryConnection':
        return { type: 'relationFrom', model, property }
      case 'queryCount':
        return { type: 'relationCountFrom', model, property }
    }
  }
  return { type: field.viewType }
}
