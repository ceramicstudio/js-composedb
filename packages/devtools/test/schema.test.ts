/* eslint-disable @typescript-eslint/no-unsafe-assignment */

import { ImageMetadataType, profilesSchema } from '@composedb/test-schemas'
import Ajv from 'ajv/dist/2020'

import { createAbstractCompositeDefinition } from '../src'
import type { AbstractCreateModelDefinition, AbstractModelDefinition } from '../src/schema/types'

function getSchema(modelDefinition: AbstractModelDefinition) {
  return (modelDefinition as AbstractCreateModelDefinition).model.schema
}

describe('schema', () => {
  it("createAbstractCompositeDefinition throws when there's no top-level model object", () => {
    expect(() => {
      createAbstractCompositeDefinition(ImageMetadataType)
    }).toThrow('No models found in Composite Definition Schema')
  })

  it("createAbstractCompositeDefinition doesn't parse unions", () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      union IntOrString = Int | String

      type ModelWithUnionProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with a property that is a union of string and int"
      ) {
        intOrStringValue: IntOrString
      }
      `)
    }).toThrow('GraphQL unions are not supported')
  })

  it("createAbstractCompositeDefinition doesn't parse interfaces", () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      interface GenericProfile {
        name: String @string(maxLength: 150)
      }
       
      type SocialProfile implements GenericProfile @createModel(
        accountRelation: SINGLE,
        description: "A model to store properties that accounts would like to share on social media"
      ) {
        description: String @string(maxLength: 420)
        emoji: String @string(maxLength: 2)
        url: String @string(maxLength: 240)
      }
      `)
    }).toThrow('GraphQL interfaces are not supported')
  })

  it('createAbstractCompositeDefinition creates an InternalCompositeDefinition for profiles from schema', () => {
    expect(createAbstractCompositeDefinition(profilesSchema)).toMatchSnapshot()
  })

  it('createAbstractCompositeDefinition creates models whose schemas conform to JSON Schema standard', () => {
    const { models } = createAbstractCompositeDefinition(profilesSchema)
    const validator = new Ajv()
    expect(validator.validateSchema(getSchema(models.GenericProfile), true)).toBe(true)
    expect(validator.validateSchema(getSchema(models.SocialProfile), true)).toBe(true)
    expect(validator.validateSchema(getSchema(models.PersonProfile), true)).toBe(true)
  })

  it('DID scalar is supported and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithDIDProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with DID properties"
      ) {
        didValue: DID
        requiredDidValue: DID!
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithDIDProp: {
          action: 'create',
          model: {
            name: 'ModelWithDIDProp',
            accountRelation: { type: 'single' },
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                didValue: { $ref: '#/$defs/GraphQLDID' },
                requiredDidValue: { $ref: '#/$defs/GraphQLDID' },
              },
              $defs: {
                GraphQLDID: {
                  type: 'string',
                  title: 'GraphQLDID',
                  pattern:
                    "^did:[a-zA-Z0-9.!#$%&'*+\\/=?^_`{|}~-]+:[a-zA-Z0-9.!#$%&'*+\\/=?^_`{|}~-]+$",
                  maxLength: 100,
                },
              },
              additionalProperties: false,
              required: ['requiredDidValue'],
            },
          },
        },
      },
    })
  })

  it('@documentAccount is only valid for DIDs', () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      type ModelWithInvalidDocumentAccountProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with an invalid @documentAccount directive"
      ) {
        nonDIDValue: String @documentAccount
      }
      `)
    }).toThrow(
      'Unsupported @documentAccount directive on field nonDIDValue of object ModelWithInvalidDocumentAccountProp, @documentAccount can only be set on a DID scalar'
    )
  })

  it('fields annotated with @documentAccount are not added to the resulting schema', () => {
    const { models } = createAbstractCompositeDefinition(`
      type ModelWithDocumentVersionProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with a @documentAccount directive"
      ) {
        floatProp: Float!
        didProp: DID @documentAccount
      }
      `)

    expect(Object.keys(models)).toHaveLength(1)
    const properties = getSchema(models.ModelWithDocumentVersionProp).properties ?? {}
    expect(properties).not.toBeFalsy()
    expect(Object.keys(properties)).toHaveLength(1)
    expect(Object.keys(properties)[0]).toEqual('floatProp')
  })

  it('CommitID scalar is supported and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithCommitIDProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with stream reference properties"
      ) {
        commitIDValue: CommitID
        requiredCommitIDValue: CommitID!
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithCommitIDProp: {
          action: 'create',
          model: {
            name: 'ModelWithCommitIDProp',
            accountRelation: { type: 'single' },
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              $defs: {
                CeramicCommitID: {
                  type: 'string',
                  title: 'CeramicCommitID',
                  maxLength: 200,
                },
              },
              properties: {
                commitIDValue: { $ref: '#/$defs/CeramicCommitID' },
                requiredCommitIDValue: { $ref: '#/$defs/CeramicCommitID' },
              },
              additionalProperties: false,
              required: ['requiredCommitIDValue'],
            },
          },
        },
      },
    })
  })

  it('@documentVersion is only valid for CommitIDs', () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      type ModelWithInvalidDocumentVersionProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with an invalid @documentVersion directive"
      ) {
        nonCommitIDValue: Int @documentVersion
      }
      `)
    }).toThrow(
      'Unsupported @documentVersion directive on field nonCommitIDValue of object ModelWithInvalidDocumentVersionProp, @documentVersion can only be set on a CommitID scalar'
    )
  })

  it('fields annotated with @documentVersion are not added to the resulting schema', () => {
    const { models } = createAbstractCompositeDefinition(`
      type ModelWithDocumentVersionProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with a @documentVersion directive"
      ) {
        numberProp: Int!
        version: CommitID @documentVersion
      }
      `)
    expect(Object.keys(models)).toHaveLength(1)
    const properties = getSchema(models.ModelWithDocumentVersionProp).properties ?? {}
    expect(Object.keys(properties)).toHaveLength(1)
    expect(Object.keys(properties)[0]).toEqual('numberProp')
  })

  it('Boolean scalar is supported and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithBooleanProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with boolean properties"
      ) {
        booleanValue: Boolean
        requiredBooleanValue: Boolean!
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithBooleanProp: {
          action: 'create',
          model: {
            name: 'ModelWithBooleanProp',
            accountRelation: { type: 'single' },
            description: 'Test model with boolean properties',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                booleanValue: {
                  type: 'boolean',
                },
                requiredBooleanValue: {
                  type: 'boolean',
                },
              },
              additionalProperties: false,
              required: ['requiredBooleanValue'],
            },
          },
        },
      },
    })
  })

  it('Int scalar is supported and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithIntProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with int properties"
      ) {
        intValue: Int
        requiredIntValue: Int!
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithIntProp: {
          action: 'create',
          model: {
            name: 'ModelWithIntProp',
            accountRelation: { type: 'single' },
            description: 'Test model with int properties',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                intValue: {
                  type: 'integer',
                },
                requiredIntValue: {
                  type: 'integer',
                },
              },
              additionalProperties: false,
              required: ['requiredIntValue'],
            },
          },
        },
      },
    })
  })

  it('Float scalar is supported and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithFloatProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with float properties"
      ) {
        floatValue: Float
        requiredFloatValue: Float!
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithFloatProp: {
          action: 'create',
          model: {
            name: 'ModelWithFloatProp',
            accountRelation: { type: 'single' },
            description: 'Test model with float properties',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                floatValue: {
                  type: 'number',
                },
                requiredFloatValue: {
                  type: 'number',
                },
              },
              additionalProperties: false,
              required: ['requiredFloatValue'],
            },
          },
        },
      },
    })
  })

  it('String scalar is supported and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithStringProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with string properties"
      ) {
        stringValue: String @string(maxLength: 3)
        requiredStringValue: String! @string(maxLength: 3)
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithStringProp: {
          action: 'create',
          model: {
            name: 'ModelWithStringProp',
            accountRelation: { type: 'single' },
            description: 'Test model with string properties',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                stringValue: {
                  type: 'string',
                },
                requiredStringValue: {
                  type: 'string',
                },
              },
              additionalProperties: false,
              required: ['requiredStringValue'],
            },
          },
        },
      },
    })
  })

  it('@string is required for Strings', () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      type ModelWithStringPropWithoutLengthDirective @createModel(
        accountRelation: SINGLE,
        description: "Test model with string property without @string directive"
      ) {
        stringValue: String
      }
      `)
    }).toThrow(
      'Missing @string directive on string field stringValue of object ModelWithStringPropWithoutLengthDirective'
    )
  })

  it('@string is required for arrays Strings', () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      type ModelWithStringArrayPropWithoutLengthDirective @createModel(
        accountRelation: SINGLE,
        description: "Test model with string array property without @length directive"
      ) {
        stringArrayValue: [String] @list(maxLength: 1)
      }
      `)
    }).toThrow(
      'Missing @string directive on string field stringArrayValue of object ModelWithStringArrayPropWithoutLengthDirective'
    )
  })

  it('ID scalar is supported and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithIDProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with GraphQL ID property"
      ) {
        idValue: ID @string(maxLength: 30)
        requiredIdValue: ID! @string(maxLength: 50)
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithIDProp: {
          action: 'create',
          model: {
            name: 'ModelWithIDProp',
            accountRelation: { type: 'single' },
            description: 'Test model with GraphQL ID property',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                idValue: {
                  type: 'string',
                  title: 'GraphQLID',
                  maxLength: 30,
                },
                requiredIdValue: {
                  type: 'string',
                  title: 'GraphQLID',
                  maxLength: 50,
                },
              },
              additionalProperties: false,
              required: ['requiredIdValue'],
            },
          },
        },
      },
    })
  })

  it('Arrays are supported and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithArrayProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with list property"
      ) {
        arrayValue: [Int] @list(maxLength: 3)
        requiredArrayValue: [Int]! @list(maxLength: 5)
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithArrayProp: {
          action: 'create',
          model: {
            name: 'ModelWithArrayProp',
            accountRelation: { type: 'single' },
            description: 'Test model with list property',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                arrayValue: {
                  type: 'array',
                  items: {
                    type: 'integer',
                  },
                  maxItems: 3,
                },
                requiredArrayValue: {
                  type: 'array',
                  items: {
                    type: 'integer',
                  },
                  maxItems: 5,
                },
              },
              additionalProperties: false,
              required: ['requiredArrayValue'],
            },
          },
        },
      },
    })
  })

  it('@string(minLength: Int, maxLength: Int!) directive is supported for strings and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithStringProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with a constrained string property"
      ) {
        stringValue: String @string(minLength: 1, maxLength: 140)
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithStringProp: {
          action: 'create',
          model: {
            name: 'ModelWithStringProp',
            accountRelation: { type: 'single' },
            description: 'Test model with a constrained string property',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                stringValue: {
                  type: 'string',
                  minLength: 1,
                  maxLength: 140,
                },
              },
              additionalProperties: false,
            },
          },
        },
      },
    })
  })

  it('@int(min: Int, max: Int) can be applied to Int, Int! or [Int] properties', () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      type ModelWithStringProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with incorrectly constrained string property"
      ) {
        intValue: String @int(min: 1) @string(maxLength: 3)
      }
      `)
    }).toThrow(
      'Unexpected @int directive with type String on field intValue of object ModelWithStringProp'
    )
  })

  it('@float(min: Float, max: Float) can be applied to Float, Float! or [Float] properties', () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      type ModelWithIntProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with incorrectly constrained int property"
      ) {
        intValue: Int @float(max: 1)
      }
      `)
    }).toThrow(
      'Unexpected @float directive with type Int on field intValue of object ModelWithIntProp'
    )
  })

  it('@string(minLength: Int, maxLength: Int!) can be applied to strings or arrays of strings', () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      type ModelWithArrayProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with incorrectly constrained array property"
      ) {
        intValue: [Int] @string(minLength: 10, maxLength: 140)
      }
      `)
    }).toThrow('Missing @list directive on list field intValue of object ModelWithArrayProp')
  })

  it('@list(minLength: Int, maxLength: Int!) can be applied to lists', () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      type ModelWithStringProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with incorrectly constrained strings property"
      ) {
        intValue: String @list(maxLength: 140) @string(maxLength: 10)
      }
      `)
    }).toThrow('Unexpected @list directive on field intValue of object ModelWithStringProp')
  })

  it('@list(minLength: Int, maxLength: Int!) directive is supported for arrays and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithArrayProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with a constrained array property"
      ) {
        arrayValue: [String] @string(maxLength: 5) @list(minLength: 10, maxLength: 15)
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithArrayProp: {
          action: 'create',
          model: {
            name: 'ModelWithArrayProp',
            accountRelation: { type: 'single' },
            description: 'Test model with a constrained array property',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                arrayValue: {
                  type: 'array',
                  minItems: 10,
                  maxItems: 15,
                  items: {
                    type: 'string',
                    maxLength: 5,
                  },
                },
              },
              additionalProperties: false,
            },
          },
        },
      },
    })
  })

  it('@string(minLength: Int, maxLength: Int!) directive is supported for arrays of strings and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithArrayProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with an array property with constrained items"
      ) {
        arrayValue: [String] @string(minLength: 4, maxLength: 440) @list(maxLength: 5)
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithArrayProp: {
          action: 'create',
          model: {
            name: 'ModelWithArrayProp',
            accountRelation: { type: 'single' },
            description: 'Test model with an array property with constrained items',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                arrayValue: {
                  type: 'array',
                  items: {
                    type: 'string',
                    minLength: 4,
                    maxLength: 440,
                  },
                },
              },
              additionalProperties: false,
            },
          },
        },
      },
    })
  })

  it('@list is required for arrays', () => {
    expect(() => {
      createAbstractCompositeDefinition(`
      type ModelWithArrayPropWithoutArrayLengthDirective @createModel(
        accountRelation: SINGLE,
        description: "Test model with an array property without @arrayLength directive"
      ) {
        arrayValue: [Int]
      }
      `)
    }).toThrow(
      'Missing @list directive on list field arrayValue of object ModelWithArrayPropWithoutArrayLengthDirective'
    )
  })

  it('@int(min: Int, max: Int) directive is supported and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithIntProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with a constreained int property"
      ) {
        intValue: Int @int(min: 5, max: 10)
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithIntProp: {
          action: 'create',
          model: {
            name: 'ModelWithIntProp',
            accountRelation: { type: 'single' },
            description: 'Test model with a constreained int property',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                intValue: {
                  type: 'integer',
                  minimum: 5,
                  maximum: 10,
                },
              },
              additionalProperties: false,
            },
          },
        },
      },
    })
  })

  it('@float(min: Float, max: Float) directive is supported and properly converted to ICD', () => {
    expect(
      createAbstractCompositeDefinition(`
      type ModelWithFloatProp @createModel(
        accountRelation: SINGLE,
        description: "Test model with a constrained float property"
      ) {
        floatValue: Float @float(min: 5.0, max: 10.0)
      }
      `)
    ).toMatchObject({
      models: {
        ModelWithFloatProp: {
          action: 'create',
          model: {
            name: 'ModelWithFloatProp',
            accountRelation: { type: 'single' },
            description: 'Test model with a constrained float property',
            schema: {
              $schema: 'https://json-schema.org/draft/2020-12/schema',
              type: 'object',
              properties: {
                floatValue: {
                  type: 'number',
                  minimum: 5.0,
                  maximum: 10.0,
                },
              },
              additionalProperties: false,
            },
          },
        },
      },
    })
  })
})
