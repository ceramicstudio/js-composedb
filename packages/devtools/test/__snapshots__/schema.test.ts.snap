// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema createAbstractCompositeDefinition creates an InternalCompositeDefinition for profiles from schema 1`] = `
{
  "commonEmbeds": [
    "ImageMetadata",
    "ImageSources",
  ],
  "models": {
    "GenericProfile": {
      "action": "create",
      "model": {
        "accountRelation": {
          "type": "single",
        },
        "description": "A model to store common profile-related properties",
        "name": "GenericProfile",
        "relations": {},
        "schema": {
          "$defs": {
            "ImageMetadata": {
              "additionalProperties": false,
              "properties": {
                "height": {
                  "minimum": 1,
                  "type": "integer",
                },
                "mimeType": {
                  "maxLength": 50,
                  "type": "string",
                },
                "size": {
                  "minimum": 1,
                  "type": "integer",
                },
                "src": {
                  "maxLength": 150,
                  "type": "string",
                },
                "width": {
                  "minimum": 1,
                  "type": "integer",
                },
              },
              "required": [
                "src",
                "mimeType",
                "width",
                "height",
              ],
              "title": "ImageMetadata",
              "type": "object",
            },
            "ImageSources": {
              "additionalProperties": false,
              "properties": {
                "alternatives": {
                  "items": {
                    "$ref": "#/$defs/ImageMetadata",
                  },
                  "maxItems": 20,
                  "type": "array",
                },
                "original": {
                  "$ref": "#/$defs/ImageMetadata",
                },
              },
              "required": [
                "original",
              ],
              "title": "ImageSources",
              "type": "object",
            },
          },
          "$schema": "https://json-schema.org/draft/2020-12/schema",
          "additionalProperties": false,
          "properties": {
            "image": {
              "$ref": "#/$defs/ImageSources",
            },
            "name": {
              "maxLength": 150,
              "type": "string",
            },
          },
          "type": "object",
        },
        "version": "1.0",
        "views": {},
      },
    },
    "PersonProfile": {
      "action": "create",
      "model": {
        "accountRelation": {
          "type": "single",
        },
        "description": "A model to store accounts' personal data",
        "name": "PersonProfile",
        "relations": {},
        "schema": {
          "$defs": {
            "GraphQLCountryCode": {
              "maxLength": 2,
              "pattern": "^(AD|AE|AF|AG|AI|AL|AM|AO|AQ|AR|AS|AT|AU|AW|AX|AZ|BA|BB|BD|BE|BF|BG|BH|BI|BJ|BL|BM|BN|BO|BQ|BR|BS|BT|BV|BW|BY|BZ|CA|CC|CD|CF|CG|CH|CI|CK|CL|CM|CN|CO|CR|CU|CV|CW|CX|CY|CZ|DE|DJ|DK|DM|DO|DZ|EC|EE|EG|EH|ER|ES|ET|FI|FJ|FK|FM|FO|FR|GA|GB|GD|GE|GF|GG|GH|GI|GL|GM|GN|GP|GQ|GR|GS|GT|GU|GW|GY|HK|HM|HN|HR|HT|HU|ID|IE|IL|IM|IN|IO|IQ|IR|IS|IT|JE|JM|JO|JP|KE|KG|KH|KI|KM|KN|KP|KR|KW|KY|KZ|LA|LB|LC|LI|LK|LR|LS|LT|LU|LV|LY|MA|MC|MD|ME|MF|MG|MH|MK|ML|MM|MN|MO|MP|MQ|MR|MS|MT|MU|MV|MW|MX|MY|MZ|NA|NC|NE|NF|NG|NI|NL|NO|NP|NR|NU|NZ|OM|PA|PE|PF|PG|PH|PK|PL|PM|PN|PR|PS|PT|PW|PY|QA|RE|RO|RS|RU|RW|SA|SB|SC|SD|SE|SG|SH|SI|SJ|SK|SL|SM|SN|SO|SR|SS|ST|SV|SX|SY|SZ|TC|TD|TF|TG|TH|TJ|TK|TL|TM|TN|TO|TR|TT|TV|TW|TZ|UA|UG|UM|US|UY|UZ|VA|VC|VE|VG|VI|VN|VU|WF|WS|YE|YT|ZA|ZM|ZW)$",
              "title": "GraphQLCountryCode",
              "type": "string",
            },
            "GraphQLDate": {
              "format": "date",
              "maxLength": 100,
              "title": "GraphQLDate",
              "type": "string",
            },
          },
          "$schema": "https://json-schema.org/draft/2020-12/schema",
          "additionalProperties": false,
          "properties": {
            "affiliations": {
              "items": {
                "maxLength": 140,
                "type": "string",
              },
              "maxItems": 20,
              "type": "array",
            },
            "birthDate": {
              "$ref": "#/$defs/GraphQLDate",
            },
            "gender": {
              "maxLength": 42,
              "type": "string",
            },
            "homeLocation": {
              "maxLength": 140,
              "type": "string",
            },
            "nationalities": {
              "items": {
                "$ref": "#/$defs/GraphQLCountryCode",
              },
              "maxItems": 5,
              "type": "array",
            },
            "residenceCountry": {
              "$ref": "#/$defs/GraphQLCountryCode",
            },
          },
          "type": "object",
        },
        "version": "1.0",
        "views": {},
      },
    },
    "SocialProfile": {
      "action": "create",
      "model": {
        "accountRelation": {
          "type": "single",
        },
        "description": "A model to store properties that accounts would like to share on social media",
        "name": "SocialProfile",
        "relations": {},
        "schema": {
          "$defs": {
            "ImageMetadata": {
              "additionalProperties": false,
              "properties": {
                "height": {
                  "minimum": 1,
                  "type": "integer",
                },
                "mimeType": {
                  "maxLength": 50,
                  "type": "string",
                },
                "size": {
                  "minimum": 1,
                  "type": "integer",
                },
                "src": {
                  "maxLength": 150,
                  "type": "string",
                },
                "width": {
                  "minimum": 1,
                  "type": "integer",
                },
              },
              "required": [
                "src",
                "mimeType",
                "width",
                "height",
              ],
              "title": "ImageMetadata",
              "type": "object",
            },
            "ImageSources": {
              "additionalProperties": false,
              "properties": {
                "alternatives": {
                  "items": {
                    "$ref": "#/$defs/ImageMetadata",
                  },
                  "maxItems": 20,
                  "type": "array",
                },
                "original": {
                  "$ref": "#/$defs/ImageMetadata",
                },
              },
              "required": [
                "original",
              ],
              "title": "ImageSources",
              "type": "object",
            },
          },
          "$schema": "https://json-schema.org/draft/2020-12/schema",
          "additionalProperties": false,
          "properties": {
            "background": {
              "$ref": "#/$defs/ImageSources",
            },
            "description": {
              "maxLength": 420,
              "type": "string",
            },
            "emoji": {
              "maxLength": 2,
              "type": "string",
            },
            "url": {
              "maxLength": 240,
              "type": "string",
            },
          },
          "type": "object",
        },
        "version": "1.0",
        "views": {},
      },
    },
  },
}
`;
